### PR TITLE
Cryptography quickstart for Go

### LINKED ISSUE #828: Cryptography quickstart
## Describe the proposal

Cryptography quickstart needed for 1.11 release

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Fixes #828

Note: there's no HTTP API example because users are **strongly** encouraged to use the SDK (gRPC-based).

### CODE REVIEW

**Comment by ItalyPaleAle** [comment]
CI is currently failing due to broken link, because the docs haven't been published yet:

image

File: cryptography/go/sdk/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Dapr secrets management (Dapr SDK)
+
+In this quickstart, you'll create an application that encrypts, and then decrypts, a short message and a large file using the Dapr cryptography APIs (high-level).
+
+Visit the documentation to learn more about the [Cryptography building block](https://docs.dapr.io/developing-applications/building-blocks/cryptography/) in Dapr.
```
Comment (ASHIQUEMD) [inline cryptography/go/sdk/README.md:5]: This link isn't available right now? is this okay?

**Review by msfussell — DISMISSED** [review]
LGTM

File: cryptography/go/sdk/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+> **Note:** This example uses the Dapr SDK. Using the Dapr SDK, which leverages gRPC internally, is **strongly** recommended when using the high-level cryptography APIs (to encrypt and decrypt messages).
```
Comment (greenie-msft) [inline cryptography/go/sdk/README.md:7]: Can we add a sentence that explains why it's recommended to use the Dapr SDKs with this API? What does it provide for me that I'd have to do myself if I wasn't using and SDK?

File: cryptography/fixtures/keys/rsa-private-key.pem
Comment (paulyuk) [inline cryptography/fixtures/keys/rsa-private-key.pem:1]: Do we have issues checking in a private key, even for a sample?

File: cryptography/go/sdk/crypto-quickstart/app.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+func main() {
+	// Create a new Dapr SDK client
+	client, err := dapr.NewClient()
+	if err != nil {
+		log.Fatalf("Failed to initialize the Dapr client: %v", err)
+	}
+	defer client.Close()
+
+	// Step 1: encrypt a message, using the RSA key
```
Comment (paulyuk) [inline cryptography/go/sdk/crypto-quickstart/app.go:32]:
```suggestion
	// Step 1: encrypt a message using the RSA key, and then decrypt to show end-to-end
```

File: cryptography/go/sdk/crypto-quickstart/app.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	encryptDecryptMessage(client)
+
+	// Step 2: encrypt a large file, using the AES key
```
Comment (paulyuk) [inline cryptography/go/sdk/crypto-quickstart/app.go:35]:
```suggestion
	// Step 2: encrypt a message using the AES key, and then decrypt to show end-to-end
```

File: cryptography/go/sdk/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# Dapr secrets management (Dapr SDK)
```
Comment (greenie-msft) [inline cryptography/go/sdk/README.md:1]: Can we also provide a little narrative about what the scenario is? What is being encrypted/decrypted?

File: cryptography/go/sdk/crypto-quickstart/app.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	encryptDecryptFile(client)
+}
+
+func encryptDecryptMessage(client dapr.Client) {
+	const message = "Ognuno sta solo sul cuor della terra\ntrafitto da un raggio di sole:\ned è subito sera."
```
Comment (paulyuk) [inline cryptography/go/sdk/crypto-quickstart/app.go:40]: While I really love the personal style and flavor, it threw me off seeing Italian when I was trying to understand the encryption itself and if the STOUT looked right.  Would it be possibly ok to use English just so less of the WW audience needs to use Google?  I'm a believer that a quickstart has examples and domain knowledge everyone has, so you focus instead on the new concepts like Crypto itself.

File: cryptography/go/sdk/crypto-quickstart/go.mod [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module dapr_example
+
+go 1.19
+
+require github.com/dapr/go-sdk v1.6.1-0.20230526171131-942dcb8512fb
```
Comment (paulyuk) [inline cryptography/go/sdk/crypto-quickstart/go.mod:5]: We have to remember to update this when we release final

File: secrets_management/go/http/order-processor/dapr_example
Comment (paulyuk) [inline secrets_management/go/http/order-processor/dapr_example:1]: This should be deleted right?  it's output of `go build .` I assume..

File: secrets_management/go/sdk/order-processor/dapr_example
Comment (paulyuk) [inline secrets_management/go/sdk/order-processor/dapr_example:1]: should delete this.

**Review by paulyuk — CHANGES_REQUESTED** [review]
Thank you for contributing this !  pls see comments.

File: cryptography/go/sdk/crypto-quickstart/app.go (hunk 6 above)
Comment (msfussell) [reply cryptography/go/sdk/crypto-quickstart/app.go:40]: Yes, We should not use Italian here. English is the international standard, including in China.

Comment (msfussell) [reply cryptography/go/sdk/crypto-quickstart/app.go:40]: Yes, We should not use Italian here. English is the international standard, including in China.

Comment (msfussell) [reply cryptography/go/sdk/crypto-quickstart/app.go:40]: +1 to not having Italian here. English is the international that most countries accept and we need to keep a consistency in the examples. Please change to english

**Comment by paulyuk** [comment]
@ItalyPaleAle  could we also please add these to .gitignore so they won't be added with commits?

image

**Review by tmacam — COMMENTED** [review]
I think we can get rid of the hardcoded RSA private key and encryption keys.

My suggestion is to have this auto-generated by the Go code itself. Keys would be written in the `cryptography/go/sdk/crypto-quickstart/keys` directory (which needs to be created). We could move `cryptography/fixtures/README.md` to this new location so it would not be an empty directory and problematic for git.

File: cryptography/go/sdk/README.md (hunk 5 above)
Comment (ItalyPaleAle) [reply cryptography/go/sdk/README.md:1]: Done

File: cryptography/go/sdk/README.md (hunk 2 above)
Comment (ItalyPaleAle) [reply cryptography/go/sdk/README.md:7]: I have checked other quickstarts and I don't see this mentioned anywhere else. The truth is that implementing this by hand is quite complex as it requires dealing with chunking.

File: secrets_management/go/http/order-processor/dapr_example
Comment (ItalyPaleAle) [reply secrets_management/go/http/order-processor/dapr_example:1]: Yes I had deleted it. It should have never been included

File: secrets_management/go/sdk/order-processor/dapr_example
Comment (ItalyPaleAle) [reply secrets_management/go/sdk/order-processor/dapr_example:1]: It's here because I deleted it.

**Comment by ItalyPaleAle** [comment]
@msfussell @paulyuk I have made the requested changes

File: cryptography/go/sdk/makefile
Comment (tmacam) [inline cryptography/go/sdk/makefile:1]: Should this also remove the keys?

Comment (ItalyPaleAle) [reply cryptography/go/sdk/makefile:1]: It does!

File: cryptography/go/sdk/README.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
+# Dapr cryptography (Dapr SDK)
+
+In this quickstart, you'll create an application that encrypts, and then decrypts, data using the Dapr cryptography APIs (high-level). We will:
+
+- Encrypt and then decrypt a short string, reading the result in-memory, in a Go byte slice
+- Encrypt and then decrypt a large file, storing the encrypted and decrypted data to files using streams
+
+Visit the documentation to learn more about the [Cryptography building block](https://docs.dapr.io/developing-applications/building-blocks/cryptography/) in Dapr.
```
Comment (paulyuk) [inline cryptography/go/sdk/README.md:8]:
```suggestion
Visit the documentation to learn more about the [Cryptography building block](https://v1-11.docs.dapr.io/developing-applications/building-blocks/cryptography/) in Dapr.
```

Comment (paulyuk) [reply cryptography/go/sdk/README.md:8]: If we don't set to 1.11 release this will 404 and fail automated tests.

**Review by paulyuk — APPROVED** [review]
LGTM with doc change.  Thank you for contribution!

**Review by yaron2 — APPROVED** [review]

**Pull request merged**